conda/core/path_actions.py

```diff
@@ -82,27 +85,84 @@
 )
 
 
-class _Action(metaclass=ABCMeta):
+class Action:
+    """Base class for path manipulation actions, including linking, unlinking, and others.
+
+    Pre and post-transaction plugins should inherit this class to implement their
+    own verification, execution, reversing, and cleanup steps. These methods are
+    guaranteed to be called in the following order:
+
+        1. ``verify``
+        2. ``execute``
+        3. ``reverse`` (only if ``execute`` raises an exception)
+        4. ``cleanup``
+
+
+    :param transaction_context: Mapping between target prefixes and PrefixActionGroup
+        instances
+    :param target_prefix: Target prefix for the action
+    :param unlink_precs: Package records to be unlinked
+    :param link_precs: Package records to link
+    :param remove_specs: Specs to be removed
+    :param update_specs: Specs to be updated
+    :param neutered_specs: Specs to be neutered
+    """
+
     _verified = False
 
+    def __init__(
+        self,
+        transaction_context,
+        target_prefix,
+        unlink_precs,
+        link_precs,
+        remove_specs,
+        update_specs,
+        neutered_specs,
```

Comment:
```suggestion
        transaction_context: dict[str, str] | None = None,
        target_prefix: str | None = None,
        unlink_precs: Iterable[PackageRecord] | None = None,
        link_precs: Iterable[PackageRecord] | None = None,
        remove_specs: Iterable[MatchSpec] | None = None,
        update_specs: Iterable[MatchSpec] | None = None,
        neutered_specs: Iterable[MatchSpec] | None = None,
```

Let's add some nice type annotations because this will be a public class. I also think that each argument should be optional in case we want to change what we pass these objects in the future.

Reply:
Thanks for this - IIRC the calling functions don't have type annotations for these, but it's nice to have the insider knowledge for this :)